multiagenticswarm/core/prompt_parser.py

```diff
@@ -0,0 +1,204 @@
+# multiagenticswarm/core/prompt_parser.py
+import re
+import json
+import os
+from typing import Dict, Any, Optional, Callable
+from langchain_core.messages import SystemMessage, HumanMessage
+from langchain_groq import ChatGroq
+from dotenv import load_dotenv
+
+# Load environment variables
+load_dotenv()
+
+# Import schema + role mappings
+from prompt_schemas import DEFAULT_FUNCTION_SCHEMA, ROLE_MAPPINGS
+print(DEFAULT_FUNCTION_SCHEMA)
+print(ROLE_MAPPINGS)
+
+# FUNCTION CALL PROMPT BUILDER
+def build_enhanced_function_call_prompt(prompt: str):
+    system_content = f"""You are an expert workflow parser for multi-agent collaboration systems.
+
+Your task is to extract structured workflow information from natural language prompts.
+
+IMPORTANT GUIDELINES:
+1. **Phases**: Identify collaboration patterns (parallel, sequential, conditional, loop)
+2. **Duration**: Use meaningful durations like "until_complete", not generic times
+3. **Rules**: Extract conditional logic, feedback loops, and validation rules
+4. **Dependencies**: Map which agents depend on others' completion
+5. **Roles**: Infer agent responsibilities from context
+
+PATTERN EXAMPLES:
+- "A and B work in parallel" → pattern: "parallel"
+- "then C reviews" → pattern: "sequential"
+- "if issues found, return to relevant agent" → rule: conditional_loop
+- "QA reviews both outputs" → dependencies: QA depends on previous agents
+
+SCHEMA TO FOLLOW:
+{json.dumps(DEFAULT_FUNCTION_SCHEMA['parameters'], indent=2)}
+
+EXAMPLE INPUT: "UI and Backend agents work in parallel, then QA agent reviews both outputs. If issues found, relevant agent fixes them."
+
+EXAMPLE OUTPUT:
+{{
+  "phases": [
+    {{
+      "pattern": "parallel",
+      "agents": ["UI", "Backend"],
+      "duration": "until_complete",
+      "conditions": []
+    }},
+    {{
+      "pattern": "sequential",
+      "agents": ["QA"],
+      "duration": "until_complete",
+      "conditions": ["all_previous_complete"]
+    }}
+  ],
+  "rules": [
+    {{
+      "type": "conditional_loop",
+      "condition": "issues_found",
+      "action": "return_to_relevant_agent"
+    }}
+  ],
+  "dependencies": {{
+    "QA": ["UI", "Backend"]
+  }},
+  "roles": {{
+    "UI": "Interface development",
+    "Backend": "API and data processing",
+    "QA": "Quality validation and testing"
+  }}
+}}
+
+Return ONLY valid JSON matching the schema. No explanations."""
+
+    return [
+        SystemMessage(content=system_content),
+        HumanMessage(content=f"Parse this collaboration prompt: {prompt}")
+    ]
+
+#ENHANCED HEURISTIC PARSER
+def heuristic_parse(prompt: str) -> Dict[str, Any]:
+    """Enhanced fallback parser with regex + role mappings"""
+    agents = re.findall(r'\b([A-Z][A-Za-z]*)\s+(?:agent|team|developer)', prompt, re.IGNORECASE)
+    if not agents:
+        agents = re.findall(r'\b(frontend|backend|UI|QA|testing|database|api|security)\b', prompt, re.IGNORECASE)
+
```

Comment:
This completely violates the "natural language → structure" requirement. It only recognizes specific agent types.